dev/InfoBar/InfoBar.cpp

```diff
@@ -0,0 +1,210 @@
+    {
+        m_closeButtonClickRevoker = closeButton.Click(winrt::auto_revoke, { this, &InfoBar::OnCloseButtonClick });
+
+        // Do localization for the close button
+        if (winrt::AutomationProperties::GetName(closeButton).empty())
+        {
+            const auto closeButtonName = ResourceAccessor::GetLocalizedStringResource(SR_InfoBarCloseButtonName);
+            winrt::AutomationProperties::SetName(closeButton, closeButtonName);
+        }
+    }
+
+    UpdateVisibility(m_notifyOpen, true);
+    m_notifyOpen = false;
+
+    UpdateSeverity();
+    UpdateIcon();
+    UpdateIconVisibility();
+    UpdateCloseButton();
+}
+
+void InfoBar::OnCloseButtonClick(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    m_closeButtonClickEventSource(*this, nullptr);
+    m_lastCloseReason = winrt::InfoBarCloseReason::CloseButton;
+    IsOpen(false);
+}
+
+void InfoBar::RaiseClosingEvent()
+{
+    auto const args = winrt::make_self<InfoBarClosingEventArgs>();
+    args->Reason(m_lastCloseReason);
+
+    m_closingEventSource(*this, *args);
+
+    if (!args->Cancel())
+    {
+        UpdateVisibility();
+        RaiseClosedEvent();
+    }
+    else
+    {
+        // The developer has changed the Cancel property to true,
+        // so we need to revert the IsOpen property to true.
+        IsOpen(true);
+    }
+}
+
+void InfoBar::RaiseClosedEvent()
+{
+    auto const args = winrt::make_self<InfoBarClosedEventArgs>();
+    args->Reason(m_lastCloseReason);
+    m_closedEventSource(*this, *args);
+}
+
+void InfoBar::OnIsOpenPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    if (IsOpen())
+    {
+        //Reset the close reason to the default value of programmatic.
+        m_lastCloseReason = winrt::InfoBarCloseReason::Programmatic;
+
+        UpdateVisibility();
+    }
+    else
+    {
+        RaiseClosingEvent();
+    }
+}
+
+void InfoBar::OnSeverityPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateSeverity();
+}
+
+void InfoBar::OnIconSourcePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateIcon();
+    UpdateIconVisibility();
+}
+
+void InfoBar::OnIsIconVisiblePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateIconVisibility();
+}
+
+void InfoBar::OnIsClosablePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateCloseButton();
+}
+
+void InfoBar::UpdateVisibility(bool notify, bool force)
+{
+    auto const peer = winrt::FrameworkElementAutomationPeer::FromElement(*this).try_as<winrt::InfoBarAutomationPeer>();
+    if (!m_applyTemplateCalled)
+    {
+        // ApplyTemplate() hasn't been called yet but IsOpen has already been set.
+        // Since this method will be called again shortly from ApplyTemplate, we'll just wait and send a notification then.
+        m_notifyOpen = true;
```

Comment:
Could we get rid of this flag here? It looks like the only time we can enter this code path is when IsOpen was set before the template was applied. The only time this flag is used is inside ApplyTemplate to indicate whether we want to notify that IsOpen was true or not. Is there a possibility for IsOpen to be false, yet `m_notifyOpen` to be true when we are in ApplyTemplate? I'm inclined to say no, but maybe you can think of a scenario?

Reply:
It could be if ApplyTemplate was called a second time after the control was created. Admittedly seems unlikely, but...

Reply:
Wouldn't that actually raise it twice then? If you set IsOpen to true before the template is getting applied the first time, then it will raise the event correctly when the first template is being applied. If you set IsOpen to false then, and apply the template again (e.g. because you have a different style based on the message or something like that), it would raise the closing event, even though it did not actually get closed.

Reply:
But on step 4, m_notifyOpen is false so it wouldn't raise the closed event again.

Reply:
Ohhhhh overlooked that line. Never mind, yes that should work correctly. But if you apply the template the second time, wouldn't IsOpen also be false/true when m_notifyOpen is false/true?

Reply:
IsOpen could be true but m_notifyOpen should still be false. (though, I am starting to think I should probably try it to make sure :P)